src/rpdk/core/data/managed-upload-infrastructure.yaml

```diff
@@ -19,6 +19,31 @@ Resources:
             Status: Enabled
             AbortIncompleteMultipartUpload:
               DaysAfterInitiation: 1
+      VersioningConfiguration:
+        Status: Enabled
+      LoggingConfiguration:
+        DestinationBucketName: { 'Fn::Sub' : "artifact-bucket-access-logs-${AWS::AccountId}-${AWS::Region}" }
+        LogFilePrefix: ArtifactBucket
+
+  AccessLogsBucket:
+    Type: AWS::S3::Bucket
+    Properties:
+      BucketName: { 'Fn::Sub' : "artifact-bucket-access-logs-${AWS::AccountId}-${AWS::Region}" }
+      AccessControl: LogDeliveryWrite
+      BucketEncryption:
+        ServerSideEncryptionConfiguration:
+        - ServerSideEncryptionByDefault:
+            SSEAlgorithm: aws:kms
+            KMSMasterKeyID: !Ref EncryptionKey
+      LifecycleConfiguration:
+        Rules:
+          - Status: Enabled
+            ExpirationInDays: 3653
+      VersioningConfiguration:
+        Status: Enabled
+      LoggingConfiguration:
```

Comment:
Is logging on the AccessLogBucket required? If these lines can be removed, then there is no need to set the BucketName anymore, which would be a better template

Reply:
Yes, it is a critical requirement on our side. I understand what you mean regarding the template and this is the least intervening way I could think of.

Reply:
It is probably the best way to enable it. I'm not convinced that pushing this to all users of the cloudformation-cli is the way to go. Not only because the template, but also because now you've created a loop and logs will cause more logs to be written. Or like how they describe it in the S3 documentation:

> When your source bucket and target bucket are the same bucket, additional logs are created for the logs that are written to the bucket. This behavior might not be ideal for your use case because it could result in a small increase in your storage billing. In addition, the extra logs about logs might make it harder to find the log that you're looking for.

 Would enabling CloudTrail DataEvents for either bucket fulfil your requirements? That's something  you can do  outside of this template.

Adding a "DestinationBucketName"  Parameter might also be way to allow different requirements for access logs, without putting lots of options in the main template

Reply:
could you maybe explain more about your last suggestion regarding Adding a "DestinationBucketName" Parameter?

Reply:
You could add a Parameter that can be changed in the CloudFormation console. The cloudformation-cli code might need be checked to make   sure it uses UsePreviousValue for all parameters when doing that though. The template could look like this

```
# [...]
Parameters:
  DestinationBucketName:
    Description: Set this if you want to write access logs to a bucket you manage
    Type: String
    Default: ""
Conditions:
  CreateLogBucket: !Not [!Equals [!Ref DestinationBucketName,  ""]]
Resources:
  ArtifactBucket:
    Type: AWS::S3::Bucket
    Properties:
      # [...]
      LoggingConfiguration:
        DestinationBucketName: !If [CreateLogBucket, !Ref AccessLogsBucket, !Ref DestinationBucketName]
        LogFilePrefix: ArtifactBucket

  AccessLogsBucket:
    Condition: CreateLogBucket
    Type: AWS::S3::Bucket
    # [..]
# [...]
```
(Not shown: removing the BucketName and LoggingConfiguration from the AccessLogsBucket too).

Reply:
I see. would I then need to refactor this part https://github.com/aws-cloudformation/cloudformation-cli/blob/fd8f1dc65633db9c910efac96dc9f466f557c1b0/src/rpdk/core/upload.py#L26
and pass the parameter there?

Reply:
I propose to not override the default with the cloudformation cli, so I wouldn't refactor to pass a parameter (but  maybe it should be something you can pass to the cli, @rjlohan will have a more  informed opinion on that).

Unfortunately this line will throw errors if there is a parameter in the template and you do not specify it when you update. https://github.xcom/aws-cloudformation/cloudformation-cli/blob/fd8f1dc65633db9c910efac96dc9f466f557c1b0/src/rpdk/core/upload.py#L98

Adding this to that api call should solve that
```
parameters=[{"ParameterKey": "DestinationBucketName", "UsePreviousValue": True}]
```

Reply:
The approach being logging in the same bucket